### PR TITLE
[InstCombine] Fold `(icmp eq/ne (or (select cond, 0/NZ, 0/NZ), X), 0)`

### LINKED ISSUE #487: Update diff April 9th 2024, 7:28:23 pm
from: https://github.com/llvm/llvm-project/commit/d3016aa889ac12fd8a047d68fed2ddaca107b990
to: https://github.com/llvm/llvm-project/commit/e248f0df14e407b8ae98cd31fb2e77fc058f3c7e
commit: 8c96b760e522002138f848af3de799051098f917

## Change Logs
from d3016aa889ac12fd8a047d68fed2ddaca107b990 to e248f0df14e407b8ae98cd31fb2e77fc058f3c7e

[e248f0df14e407b8ae98cd31fb2e77fc058f3c7e](https://github.com/llvm/llvm-project/commit/e248f0df14e407b8ae98cd31fb2e77fc058f3c7e) [Flang] Update Extensions.md for supported BIND(C) LOGICAL kind. (#88159)
[7599d478efb1576b5013d17a70971f76d6f7c25a](https://github.com/llvm/llvm-project/commit/7599d478efb1576b5013d17a70971f76d6f7c25a) [InstCombine] Fold `(icmp eq/ne (add nuw x, y), 0)` -&gt; `(icmp eq/ne (or x, y), 0)`
[759bab068157d93a71ef20dc28a2eaed4fec6d40](https://github.com/llvm/llvm-project/commit/759bab068157d93a71ef20dc28a2eaed4fec6d40) [InstCombine] Add tests for folding `(icmp eq/ne (add nuw x, y), 0)`; NFC
[5d9d740c39a6cf21f0d54ec572aed3c2f556cbcb](https://github.com/llvm/llvm-project/commit/5d9d740c39a6cf21f0d54ec572aed3c2f556cbcb) Remove the unused IntervalPartition analysis pass (#88133)

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstCombine] Add tests for folding `(icmp eq/ne (or (select cond, 0/NZ, 0/NZ), X), 0)`; NFC**
- **[InstCombine] Fold `(icmp eq/ne (or (select cond, 0/NZ, 0/NZ), X), 0)`**

Four cases:
`(icmp eq (or (select cond, 0, NonZero), Other))`
 -> `(and cond, (icmp eq Other, 0))`
`(icmp ne (or (select cond, NonZero, 0), Other))`
 -> `(or cond, (icmp ne Other, 0))`
`(icmp ne (or (select cond, 0, NonZero), Other))`
 -> `(or (not cond), (icmp ne Other, 0))`
`(icmp eq (or (select cond, NonZero, 0), Other))`
 -> `(and (not cond), (icmp eq Other, 0))`

These cases came up in tests on: https://github.com/llvm/llvm-project/pull/88088

Proofs: https://alive2.llvm.org/ce/z/ojGo_J

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/icmp-or-of-select-with-zero.ll
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/icmp-or-of-select-with-zero.ll:2]: The lack of fixed point is due to the use of `freelyInvertAllUses`.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3501]: Don't we need one-use on BO here? We're producing two instructions. I think multi-use is not tested for this fold?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3518]: Is all this complexity needed for motivating patterns? Can we only handle one-use BO + select instead?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3501]: Yeah you're right, I got fixated on the multi-use select case.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3518]: I'm happy to replace the `canFreelyInvertAllUses`/`freelyInvertAlluses` with just `getFreelyInverted`. @dtcxzyw does that cover the case you saw?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3518]: FWIW we will only be saving 5/6 LOC.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3518]: The freelyInvertAllUses() is the main thing I'd want to get rid of, yes.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3518]: @dtcxzyw does the case you had req inverting all uses or is that unnecessary?

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3518]: I prefer to create extra not instructions and add one-use constraints on BO and select.

All occurrences of this pattern in my benchmark:
```
; 77 occurrences:
; fmt/optimized/format.cc.ll
; fmt/optimized/gtest-extra.cc.ll
; fmt/optimized/header-only-test.cc.ll
; fmt/optimized/os.cc.ll
; fmt/optimized/unicode-test.cc.ll
; fmt/optimized/util.cc.ll
; lief/optimized/ASN1Reader.cpp.ll
; lief/optimized/AndroidIdent.cpp.ll
; lief/optimized/BinaryParser.cpp.ll
; lief/optimized/BuildToolVersion.cpp.ll
; lief/optimized/Builder.cpp.ll
; lief/optimized/ChainedFixup.cpp.ll
; lief/optimized/CodeIntegrity.cpp.ll
; lief/optimized/CodeView.cpp.ll
; lief/optimized/CodeViewPDB.cpp.ll
; lief/optimized/CoreAuxv.cpp.ll
; lief/optimized/CoreFile.cpp.ll
; lief/optimized/CorePrPsInfo.cpp.ll
; lief/optimized/CorePrStatus.cpp.ll
; lief/optimized/CoreSigInfo.cpp.ll
; lief/optimized/Debug.cpp.ll
; lief/optimized/DelayImport.cpp.ll
; lief/optimized/DelayImportEntry.cpp.ll
; lief/optimized/DyldChainedFixups.cpp.ll
; lief/optimized/DyldInfo.cpp.ll
; lief/optimized/DynamicEntryArray.cpp.ll
; lief/optimized/DynamicEntryLibrary.cpp.ll
; lief/optimized/DynamicEntryRpath.cpp.ll
; lief/optimized/DynamicEntryRunPath.cpp.ll
; lief/optimized/DynamicSharedObject.cpp.ll
; lief/optimized/File.cpp.ll
; lief/optimized/FileStream.cpp.ll
; lief/optimized/GenericContent.cpp.ll
; lief/optimized/LangCodeItem.cpp.ll
; lief/optimized/LinkEdit.cpp.ll
; lief/optimized/LoadConfiguration.cpp.ll
; lief/optimized/LoadConfigurationV0.cpp.ll
; lief/optimized/LoadConfigurationV1.cpp.ll
; lief/optimized/LoadConfigurationV10.cpp.ll
; lief/optimized/LoadConfigurationV11.cpp.ll
; lief/optimized/LoadConfigurationV3.cpp.ll
; lief/optimized/LoadConfigurationV4.cpp.ll
; lief/optimized/LoadConfigurationV5.cpp.ll
; lief/optimized/LoadConfigurationV6.cpp.ll
; lief/optimized/LoadConfigurationV7.cpp.ll
; lief/optimized/LoadConfigurationV8.cpp.ll
; lief/optimized/LoadConfigurationV9.cpp.ll
; lief/optimized/MemoryStream.cpp.ll
; lief/optimized/Note.cpp.ll
; lief/optimized/NoteAbi.cpp.ll
; lief/optimized/NoteGnuProperty.cpp.ll
; lief/optimized/PKCS9SigningTime.cpp.ll
; lief/optimized/Pogo.cpp.ll
; lief/optimized/PogoEntry.cpp.ll
; lief/optimized/RelocationEntry.cpp.ll
; lief/optimized/RelocationFixup.cpp.ll
; lief/optimized/Repro.cpp.ll
; lief/optimized/ResourceNode.cpp.ll
; lief/optimized/ResourceStringFileInfo.cpp.ll
; lief/optimized/ResourceVarFileInfo.cpp.ll
; lief/optimized/ResourceVersion.cpp.ll
; lief/optimized/ResourcesManager.cpp.ll
; lief/optimized/ResourcesParser.cpp.ll
; lief/optimized/RichHeader.cpp.ll
; lief/optimized/Signature.cpp.ll
; lief/optimized/SignatureParser.cpp.ll
; lief/optimized/SignerInfo.cpp.ll
; lief/optimized/SpanStream.cpp.ll
; lief/optimized/SpcIndirectData.cpp.ll
; lief/optimized/StackSize.cpp.ll
; lief/optimized/TLS.cpp.ll
; lief/optimized/TrieNode.cpp.ll
; lief/optimized/Type.cpp.ll
; lief/optimized/VectorStream.cpp.ll
; lief/optimized/hash_stream.cpp.ll
; lief/optimized/x509.cpp.ll
; spdlog/optimized/bundled_fmtlib_format.cpp.ll
define i1 @func000000000000121b(i1 %0, i32 %1, i1 %2) {
entry:
  %3 = icmp sgt i32 %1, -1
  call void @llvm.assume(i1 %3)
  %4 = zext i1 %2 to i32
  %5 = select i1 %0, i32 %1, i32 0
  %6 = or i32 %5, %4
  %7 = icmp eq i32 %6, 0
  ret i1 %7
}

; 1 occurrences:
; libquic/optimized/time_posix.cc.ll
define i1 @func0000000501007f1b(i8 %0, i64 %1, i1 %2) {
entry:
  %3 = select i1 %2, i8 2, i8 1
  %4 = icmp sgt i64 %1, -1
  %5 = select i1 %4, i8 0, i8 %3
  %6 = or i8 %5, %0
  %7 = icmp eq i8 %6, 0
  ret i1 %7
}

; 1 occurrences:
; linux/optimized/intel_hdcp.ll
define i1 @func000000000000411b(i32 %0, i1 %1) {
entry:
  %2 = select i1 %1, i32 0, i32 -110
  %3 = or i32 %2, %0
  %4 = icmp eq i32 %3, 0
  ret i1 %4
}

; 1 occurrences:
; qemu/optimized/tcg.c.ll
define i1 @func00000000051067fb(i1 %0, i32 %1) {
entry:
  %2 = lshr i32 %1, 3
  %3 = and i32 %2, 1
  call void @use.t0(i32 %3)
  %4 = select i1 %0, i32 16384, i32 0
  %5 = or disjoint i32 %3, %4
  %6 = icmp eq i32 %5, 0
  ret i1 %6
}

declare void @use.t0(i32)

; 1 occurrences:
; qemu/optimized/hw_net_cadence_gem.c.ll
define i1 @func00000000122c08db(i64 %0, i64 %1, i32 %2) {
entry:
  %3 = icmp sgt i64 %0, -1
  call void @llvm.assume(i1 %3)
  %4 = sub i32 0, %2
  %5 = and i32 %2, %4
  %6 = icmp sgt i32 %5, -1
  call void @llvm.assume(i1 %6)
  %7 = icmp eq i32 %5, 0
  %8 = select i1 %7, i64 0, i64 %1
  %9 = or disjoint i64 %8, %0
  %10 = icmp eq i64 %9, 0
  ret i1 %10
}

; 1 occurrences:
; qemu/optimized/util_host-utils.c.ll
define i1 @func000000000000841b(i64 %0, i1 noundef %1, i64 %2) {
entry:
  %3 = select i1 %1, i64 0, i64 %2
  %4 = or i64 %3, %0
  %5 = icmp eq i64 %4, 0
  ret i1 %5
}
```

I will make the dataset public this month :)

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3518]: Okay, removed all the `canfreelyinvertalluses` logic. I keep a check for `getFreelyInverted` if the sel is not single-use as thats pretty trivial.

**Comment by goldsteinn** [comment]
ping

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3497]:
```suggestion
    if (C.isZero() &&
        match(BO, m_OneUse(m_c_Or(m_Select(m_Value(Cond), m_Value(TV), m_Value(FV)),
                         m_Value(Other))))) {
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3486]:
```suggestion
    // (icmp eq (or (select cond, 0, NonZero), Other), 0)
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3493]:
```suggestion
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3511]:
```suggestion
      // avoid creating extra instructions.
    // (icmp ne (or (select cond, 0, NonZero), Other))
    //  -> (or (not cond), (icmp ne Other, 0))
    // (icmp eq (or (select cond, NonZero, 0), Other))
    //  -> (and (not cond), (icmp eq Other, 0))
```

It would be better to move the comment here?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3525]: It is incorrect. We will fallback into `getFreelyInverted` even if the opcode doesn't match :(


```suggestion
        if (match(BO, m_c_Or(m_Select(m_Specific(Cond), m_Specific(TV),
                                               m_Specific(FV)),
                             m_Value()))) {
                             
       if (BO->hasOneUse())          
         NotCond = Builder.CreateNot(Cond);
        // Otherwise, see if we can get NotCond for free.
        else
          NotCond =
              getFreelyInverted(Cond, /*WillInvertAllUses=*/false, &Builder);
      }
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3493]: I'm not seeing this diff here...

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3525]: its the `select` having one use, not the BO. But will change.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3525]: Fixed + tests. Good catch :)

**Comment by goldsteinn** [comment]
ping

**Comment by goldsteinn** [comment]
ping

**Comment by dtcxzyw** [comment]
@goldsteinn Can you have a look at the ir diff? https://github.com/dtcxzyw/llvm-opt-benchmark/pull/491
There are some unnecessary and/or insts before branches.

**Comment by goldsteinn** [comment]
> @goldsteinn Can you have a look at the ir diff? [dtcxzyw/llvm-opt-benchmark#491](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/491) There are some unnecessary and/or insts before branches.

Sorry for taking so long to get to this... can you re-run, I went through a few of them but they seem to have disappeared...

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3506]:
```suggestion
              ? (match(TV, m_Zero()) && isKnownNonZero(FV, Q))
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3501]:
```suggestion
              m_OneUse(m_c_Or(m_CombineAnd(m_Instruction(SelInst), m_Select(m_Value(Cond), m_Value(TV), m_Value(FV))),
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3529]: Use `SelInst` to avoid matching twice.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3533]: Which test covers this case?

**Comment by dtcxzyw** [comment]
> > @goldsteinn Can you have a look at the ir diff? [dtcxzyw/llvm-opt-benchmark#491](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/491) There are some unnecessary and/or insts before branches.
> 
> Sorry for taking so long to get to this... can you re-run, I went through a few of them but they seem to have disappeared...

Yeah, IR diff looks good now :)

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3533]: Yeah dropping, it doesnt make sense, if `cond` was freely ivnertable we would have swapped select arms. Only case we don't is logical and/or but this fold doesn't really make sense for `i1`

File: llvm/test/Transforms/InstCombine/icmp-or-of-select-with-zero.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/icmp-or-of-select-with-zero.ll:191]:
```suggestion
```

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for approval from @nikic.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request closed without merging**
